File: back/src/main/java/com/back/global/util/AesEncryptionUtil.java
package com.back.global.util;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import javax.crypto.Cipher;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import java.util.Base64;

@Component
public class AesEncryptionUtil {

    private static final String ALGORITHM = "AES/CBC/PKCS5Padding";
    private final SecretKeySpec secretKeySpec;

    public AesEncryptionUtil(@Value("${custom.oauth.token.encryption-key}") String encryptionKey) {
        byte[] keyBytes = Base64.getDecoder().decode(encryptionKey);
        this.secretKeySpec = new SecretKeySpec(keyBytes, "AES");
    }

    public String encrypt(String plainText) {
        try {
            byte[] iv = new byte[16];
            new SecureRandom().nextBytes(iv);
            IvParameterSpec ivSpec = new IvParameterSpec(iv);

            Cipher cipher = Cipher.getInstance(ALGORITHM);
            cipher.init(Cipher.ENCRYPT_MODE, secretKeySpec, ivSpec);

            byte[] encrypted = cipher.doFinal(plainText.getBytes(StandardCharsets.UTF_8));

            byte[] combined = new byte[iv.length + encrypted.length];
            System.arraycopy(iv, 0, combined, 0, iv.length);
            System.arraycopy(encrypted, 0, combined, iv.length, encrypted.length);

            return Base64.getEncoder().encodeToString(combined);
        } catch (Exception e) {
            throw new RuntimeException("암호화 실패", e);
        }
    }

    public String decrypt(String encryptedText) {
        try {
            byte[] combined = Base64.getDecoder().decode(encryptedText);

            byte[] iv = new byte[16];
            byte[] encrypted = new byte[combined.length - 16];
            System.arraycopy(combined, 0, iv, 0, 16);
            System.arraycopy(combined, 16, encrypted, 0, encrypted.length);

            IvParameterSpec ivSpec = new IvParameterSpec(iv);
            Cipher cipher = Cipher.getInstance(ALGORITHM);
            cipher.init(Cipher.DECRYPT_MODE, secretKeySpec, ivSpec);

            return new String(cipher.doFinal(encrypted), StandardCharsets.UTF_8);
        } catch (Exception e) {
            throw new RuntimeException("복호화 실패", e);
        }
    }
}

File: back/src/main/kotlin/com/back/global/util/AesEncryptionUtil.kt
package com.back.global.util

import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.security.SecureRandom
import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec

@Component
class AesEncryptionUtil(
    @Value("\${custom.oauth.token.encryption-key}")
    encryptionKey: String,
) {
    private val secretKeySpec = SecretKeySpec(Base64.getDecoder().decode(encryptionKey), AES)

    fun encrypt(plainText: String): String {
        try {
            val iv = ByteArray(IV_LENGTH).also(secureRandom::nextBytes)
            val encrypted = Cipher.getInstance(ALGORITHM)
                .apply { init(Cipher.ENCRYPT_MODE, secretKeySpec, IvParameterSpec(iv)) }
                .doFinal(plainText.toByteArray(Charsets.UTF_8))

            return Base64.getEncoder().encodeToString(iv + encrypted)
        } catch (e: Exception) {
            throw IllegalStateException("암호화 실패", e)
        }
    }

    fun decrypt(encryptedText: String): String {
        try {
            val combined = Base64.getDecoder().decode(encryptedText)
            val iv = combined.copyOfRange(0, IV_LENGTH)
            val encrypted = combined.copyOfRange(IV_LENGTH, combined.size)

            return Cipher.getInstance(ALGORITHM)
                .apply { init(Cipher.DECRYPT_MODE, secretKeySpec, IvParameterSpec(iv)) }
                .doFinal(encrypted)
                .toString(Charsets.UTF_8)
        } catch (e: Exception) {
            throw IllegalStateException("복호화 실패", e)
        }
    }

    companion object {
        private const val AES = "AES"
        private const val ALGORITHM = "AES/CBC/PKCS5Padding"
        private const val IV_LENGTH = 16
        private val secureRandom = SecureRandom()
    }
}
